src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java

```diff
@@ -421,37 +456,42 @@ protected long fetchOffsetTokenWithRetry() {
   }
 
   /**
-   * Fallback function to execute when retries for fetching getOffsetToken for channel fails.
+   * If a valid offset is found from snowflake, we will reset the topicPartition with
+   * (offsetReturnedFromSnowflake + 1).
+   *
+   * <p>Please note, this is done only after a failure in fetching offset. Idea behind resetting
+   * offset (1 more than what we found in snowflake) is that Kafka should send offsets from this
+   * offset number so as to not miss any data.
    *
-   * <p>Fallback function is to re-open the channel and fetch the latestOffsetToken one more time
-   * after reopen is successful.
+   * @return offset which was last present in Snowflake
+   */
+  private long offsetTokenFallbackSupplier() {
```

Comment:
why we need to return the latest offset? Seems like we finish everything inside this function

Reply:
True. 
1. The caller of this function can be `precomputeOffsetTokenForChannel` or
2. preCommit
3. Both of which needs to know what is the offset present in snowflake. 
4. Also, the fallbackSupplier needs a long value back because it is a fallback. (Similar to `map.getOrDefault()`)